lib/trie/hash.go

```diff
@@ -5,195 +5,310 @@ package trie
 
 import (
 	"bytes"
-	"context"
+	"errors"
+	"fmt"
 	"hash"
+	"io"
 	"sync"
 
 	"github.com/ChainSafe/gossamer/lib/common"
 	"github.com/ChainSafe/gossamer/pkg/scale"
 	"golang.org/x/crypto/blake2b"
-	"golang.org/x/sync/errgroup"
 )
 
-// Hasher is a wrapper around a hash function
-type hasher struct {
-	hash     hash.Hash
-	tmp      bytes.Buffer
-	parallel bool // Whether to use parallel threads when hashing
+var encodingBufferPool = &sync.Pool{
+	New: func() interface{} {
+		const initialBufferCapacity = 1900000 // 1.9MB, from checking capacities at runtime
+		b := make([]byte, 0, initialBufferCapacity)
+		return bytes.NewBuffer(b)
+	},
 }
 
-// hasherPool creates a pool of Hasher.
-var hasherPool = sync.Pool{
+var digestBufferPool = &sync.Pool{
 	New: func() interface{} {
-		h, _ := blake2b.New256(nil)
-		var buf bytes.Buffer
-		// This allocation will be helpful for encoding keys. This is the min buffer size.
-		buf.Grow(700)
-
-		return &hasher{
-			tmp:  buf,
-			hash: h,
-		}
+		const bufferCapacity = 32
+		b := make([]byte, 0, bufferCapacity)
+		return bytes.NewBuffer(b)
 	},
 }
 
-// NewHasher create new Hasher instance
-func newHasher(parallel bool) *hasher {
-	h := hasherPool.Get().(*hasher)
-	h.parallel = parallel
-	return h
+var hasherPool = &sync.Pool{
+	New: func() interface{} {
+		hasher, err := blake2b.New256(nil)
+		if err != nil {
```

Comment:
Well, I agree and disagree because

1. If someone in the future decides to set a key for whatever future reason, a panic would help out quickly find why it's crashing
2. `blake2b` is still in the experimental part of the standard library

Let me know if I'm way too paranoid though :smile_cat:

Reply:
I think linking to this discussion as comment above the panic would be nice.

Reply:
This is done in #2077 (https://github.com/ChainSafe/gossamer/pull/2077/files#diff-2ab83e082ae5556c9b4a34bd55d1068ffeb45fa6212e848643fcc9f161cf77bbR36)